lib/auth/accountrecovery_test.go

```diff
@@ -111,3 +133,264 @@ func TestRecoveryCodeEventsEmitted(t *testing.T) {
+		})
+		require.Error(t, err)
+
+		if i == defaults.MaxLoginAttempts {
+			require.True(t, trace.IsAccessDenied(err))
+		}
+	}
+
+	// Test recovery is still allowed after login lock.
+	_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+		Username:     u.username,
+		RecoveryCode: []byte(u.recoveryCodes[0]),
+		RecoverType:  types.UserTokenUsage_RECOVER_2FA,
+	})
+	require.NoError(t, err)
+
+	// Trigger max failed recovery attempts.
+	for i := 1; i <= defaults.MaxAccountRecoveryAttempts; i++ {
+		_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+			Username: u.username,
+		})
+		require.Error(t, err)
+
+		// The third failed attempt should return error.
+		if i == defaults.MaxAccountRecoveryAttempts {
+			require.EqualValues(t, ErrMaxFailedRecoveryAttempts, err)
+		}
+	}
+
+	// Test recovery is denied from attempt recovery lock.
+	_, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+		Username:     u.username,
+		RecoveryCode: []byte(u.recoveryCodes[1]),
+		RecoverType:  types.UserTokenUsage_RECOVER_2FA,
+	})
+	require.True(t, trace.IsAccessDenied(err))
+}
+
+type userAuthCreds struct {
+	recoveryCodes []string
+	username      string
+	password      []byte
+	u2fKey        *mocku2f.Key
+}
+
+func createUserWithSecondFactorAndRecoveryCodes(srv *TestTLSServer, secondFactor string) (*userAuthCreds, error) {
+	ctx := context.Background()
+	username := "fake@fake.com"
+	password := []byte("abc123")
+
+	// Enable second factors.
+	ap, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+		Type:         constants.Local,
+		SecondFactor: constants.SecondFactorOn,
+		U2F: &types.U2F{
+			AppID:  "teleport",
+			Facets: []string{"teleport"},
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := srv.Auth().SetAuthPreference(ctx, ap); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	_, _, err = CreateUserAndRole(srv.Auth(), username, []string{username})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Insert recovery code for user.
+	recoveryCodes, err := srv.Auth().generateAndUpsertRecoveryCodes(ctx, username)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	resetToken, err := srv.Auth().CreateResetPasswordToken(context.TODO(), CreateUserTokenRequest{
+		Name: username,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Add second factor.
+	if secondFactor == "otp" {
+		otp, err := getOTPCode(srv, resetToken.GetName())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		err = srv.Auth().changeUserSecondFactor(&proto.ChangeUserAuthenticationRequest{
+			TokenID: resetToken.GetName(),
+			NewMFARegisterResponse: &proto.MFARegisterResponse{Response: &proto.MFARegisterResponse_TOTP{
+				TOTP: &proto.TOTPRegisterResponse{Code: otp},
+			}},
+		}, resetToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	var u2fKey *mocku2f.Key
+	if secondFactor == "u2f" {
```

Comment:
secondFactor is always "otp" - drop the if and the parameter?

Reply:
sorry i should've added it later when i will needed it, but it gets used in the next PR for testing